src/transformers/models/llama/modeling_llama.py

```diff
@@ -57,6 +69,59 @@ def _make_causal_mask(
     return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
 
 
+def _convert_to_padding_mask(attention_mask: torch.Tensor, mask_value: float = 0.0):
+    """
+    Convert causal attention mask to key-padding mask
+    """
+    if len(attention_mask.size()) != 4:
+        raise ValueError(
+            "Expecting attention_mask to have 4 dimensions, got tensor of shape: " f"{attention_mask.size()}"
+        )
+
+    batch_size = attention_mask.size(0)
+    key_length = attention_mask.size(-1)
+
+    padding_mask = torch.ones((batch_size, key_length), device=attention_mask.device)
+
+    for i in range(batch_size):
+        mask_slice = attention_mask[i, :, -1, :]
+        padding_mask[i, :] = torch.all(mask_slice == mask_value, dim=0)
+
+    return padding_mask
+
+
+def recursively_replace_module(model, old_class, target_class):
+    """
+    Recursively replace all old_class instances of the model with a target class. The target class should have the same
+    sub-module names than the old class.
+
+    Args:
+        model (`torch.nn.Module`):
+            The model or the child module used for recursion
+        old_class (`class`):
+            The target old class to replace
+        target_class (`class`):
+            The new class that is going to be used in the replaced module.
+    """
+    for name, module in model.named_children():
+        if isinstance(module, old_class):
+            torch_device = module.q_proj.weight.device
+            with torch.device(torch_device):
+                new_module = target_class(module.config)
+
+            for inner_module_name, inner_module in module.named_modules():
```

Comment:
should use [`module.named_children()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_children) instead of [`module.named_modules()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_modules)
1. `named_modules()` will return self module with blank name
2. the function is already recursively, should only check children module instead of all. (but it will have same result in llama)